aimmo-game/simulation/turn_manager.py

```diff
@@ -79,42 +39,97 @@ def run_turn(self):
 
-        for avatar in avatars:
-            self._register_action(avatar)
-            with state_provider as game_state:
-                location_to_clear = avatar.action.target_location
-                avatar.action.process(game_state.world_map)
-                game_state.world_map.clear_cell_actions(location_to_clear)
+            with self.game_state:
+                action.process(self.game_state, NO_OTHER_ACTIONS)
 
 
 class ConcurrentTurnManager(TurnManager):
     def run_turn(self):
         '''
-        Concurrently get the intended actions from all avatars and register
-        them on the world map. Then apply actions in order of priority.
+        Concurrently get the intended actions from all avatars and apply them
+        in order of priority.
         '''
-        with state_provider as game_state:
-            avatars = game_state.avatar_manager.active_avatars
-
-        threads = [Thread(target=self._register_action,
-                          args=(avatar,)) for avatar in avatars]
+        threads = [DecisionThread(avatar, self.game_state.view(avatar))
+                   for avatar in self.game_state._avatar_manager.active_avatars]
 
         [thread.start() for thread in threads]
-        [thread.join() for thread in threads]
 
-        # Waits applied first, then attacks, then moves.
-        avatars.sort(key=lambda a: PRIORITIES[type(a.action)])
+        actions = ActionRegistry()
+        for thread in threads:
+            actions.add(thread.result())
+
+        for action in actions.sorted_by_type:
+            with self.game_state:
+                action.process(self.game_state, actions)
+
+
+class DecisionThread(Thread):
+    '''
+    Thread wrapper to get an avatar's decided action.
+    '''
+    def __init__(self, avatar, state_view):
+        self._queue = Queue()
+        super(DecisionThread, self).__init__(
+            target=self.wrapper,
+            args=(avatar, state_view)
+        )
+
+    def wrapper(self, avatar, state_view):
+        self._queue.put(avatar.decide_action(state_view))
+
+    def result(self):
+        return self._queue.get()
+
+
+class ActionRegistry(object):
+    '''
+    Class to keep track of the actions that avatars intend to perform
+    in a concurrent turn.
+    '''
+    def __init__(self):
+        self.clear()
+
+    def add(self, action):
+        self._actions_by_avatar[action.avatar_id] = action
+        self._by_target(action.target).append(action)
+
+    def clear(self):
+        self._actions_by_avatar = {}
+        self._actions_by_target = {}
+
+    def by_avatar(self, avatar_id):
+        return self._actions_by_avatar.setdefault(avatar_id, None)
+
+    def avatar_moving(self, avatar_id):
+        return isinstance(self.by_avatar(avatar_id), MoveAction)
+
+    def _by_target(self, target):
+        return self._actions_by_target.setdefault(target, [])
+
+    def by_target(self, target):
+        return (action for action in self._by_target(target))
```

Comment:
Should return a "copy of the list" so that the internal structure of the registry won't be altered from outside - I use a generator in principle because of memory usage, though in this case it doesn't matter since the length of the list is at most 4.